File: server/AsanInvest.Application/AuthService.cs
using System.Security.Cryptography;
using System.Text.Json;
using AsanInvest.Domain;
using AsanInvest.Domain.Entities;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Options;

namespace AsanInvest.Application;

public sealed class AuthService
{
    private readonly IAppDbContext _db;
    private readonly AppSettings _settings;
    private readonly AuthChallengeStore _challenges;
    private readonly IEmailSender _email;

    public AuthService(IAppDbContext db, IOptions<AppSettings> settings, AuthChallengeStore challenges, IEmailSender email)
    {
        _db = db;
        _settings = settings.Value;
        _challenges = challenges;
        _email = email;
    }

    public async Task<AuthOutcome> RegisterAsync(RegisterRequest req, CancellationToken ct)
    {
        var email = req.Email.Trim().ToLowerInvariant();
        if (await _db.Users.AnyAsync(u => u.Email == email, ct))
            throw AppException.Conflict("An account with this email already exists", "EMAIL_TAKEN");

        var autoVerify = !_settings.RequireEmailVerification;
        var user = new User
        {
            Email = email,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(req.Password, _settings.BcryptRounds),
            Locale = req.Locale ?? "az",
            Status = UserStatus.ACTIVE,
            AuthProvider = AuthProvider.EMAIL,
            EmailVerifiedAt = autoVerify ? DateTimeOffset.UtcNow : null,
            IdentificationLevel = IdentificationLevel.BASIC,
            Consents = req.Consents is null ? "{}" : JsonSerializer.Serialize(req.Consents),
            ConsentVersion = req.Consents?.Version,
            ConsentedAt = req.Consents is null ? null : DateTimeOffset.UtcNow,
            Profile = new Profile(),
            RoleAssignments = { new UserRoleAssignment { Role = UserRole.INVESTOR } },
        };
        _db.Users.Add(user);
        await _db.SaveChangesAsync(ct);

        if (!autoVerify)
        {
            var verify = Tokens.SignPurpose(_settings, user.Id, "email_verify", lifetime: "24h");
            await _email.SendAsync(
                user.Email,
                "Verify your ASAN Invest email",
                $"Open {_settings.ClientOrigin}/verify-email?token={Uri.EscapeDataString(verify)} to confirm this address.",
                ct);
        }

        if (req.GuestSessionToken is { } token && Guid.TryParse(token, out var sid))
        {
            var session = await _db.GuestSessions.FirstOrDefaultAsync(s => s.Id == sid, ct);
            if (session is not null && session.ExpiresAt > DateTimeOffset.UtcNow && user.Profile is not null)
            {
                var contacts = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(user.Profile.Contacts) ?? new();
                contacts["guestAnswers"] = JsonSerializer.Deserialize<JsonElement>(session.Answers);
                user.Profile.Contacts = JsonSerializer.Serialize(contacts);
                session.ConvertedUserId = user.Id;
                await _db.SaveChangesAsync(ct);
            }
        }

        await Audit(user.Id, "auth.register", "user", user.Id.ToString(), ct);
        return new AuthOutcome { Tokens = TokenPayload(user) };
    }

    public async Task<AuthOutcome> LoginAsync(string email, string password, CancellationToken ct)
    {
        var user = await _db.Users.Include(u => u.RoleAssignments).Include(u => u.Profile)
            .FirstOrDefaultAsync(u => u.Email == email.Trim().ToLowerInvariant(), ct);
        if (user?.PasswordHash is null || !BCrypt.Net.BCrypt.Verify(password, user.PasswordHash))
            throw AppException.Unauthorized("Invalid email or password", "INVALID_CREDENTIALS");
        if (user.Status == UserStatus.DISABLED) throw AppException.Forbidden("Account is disabled", "ACCOUNT_DISABLED");
        if (user.EmailVerifiedAt is null && _settings.RequireEmailVerification)
            throw AppException.Forbidden("Email verification required", "EMAIL_NOT_VERIFIED");

        var roles = Roles.Active(user.RoleAssignments, DateTimeOffset.UtcNow);
        if (Roles.RequiresTwoFactor(roles))
        {
            var testing = string.Equals(Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT"), "Testing", StringComparison.OrdinalIgnoreCase);
            var code = testing ? "123456" : NewOtp();
            var challengeId = _challenges.IssueOtp(user.Id, Tokens.Hash(code), TimeSpan.FromMinutes(10));
            await _email.SendAsync(user.Email, "ASAN Invest sign-in code", $"Your sign-in code expires in 10 minutes. Code: {code}", ct);
            return new AuthOutcome { TwoFactorRequired = true, ChallengeId = challengeId };
        }

        await Audit(user.Id, "auth.login", "user", user.Id.ToString(), ct);
        return new AuthOutcome { Tokens = TokenPayload(user) };
    }

    public async Task<AuthOutcome> VerifyTwoFactorAsync(string challengeId, string code, CancellationToken ct)
    {
        if (!_challenges.ConsumeOtp(challengeId, Tokens.Hash(code), out var userId))
            throw AppException.Unauthorized("Two-factor code is invalid", "TWO_FACTOR_INVALID");
        var user = await _db.Users.Include(u => u.RoleAssignments).Include(u => u.Profile).FirstOrDefaultAsync(u => u.Id == userId, ct)
            ?? throw AppException.Unauthorized();
        user.TwoFactorEnabled = true;
        await _db.SaveChangesAsync(ct);
        return new AuthOutcome { Tokens = TokenPayload(user) };
    }

    public async Task<AuthOutcome> RefreshAsync(string? refreshToken, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(refreshToken)) throw AppException.Unauthorized("Refresh token missing", "REFRESH_MISSING");
        try
        {
            var jwt = Tokens.Require(refreshToken, _settings.JwtRefreshSecret, "refresh");
            var jti = jwt.Id;
            if (_challenges.IsRefreshRevoked(jti)) throw AppException.Unauthorized("Refresh token is invalid", "REFRESH_INVALID");
            var user = await _db.Users.Include(u => u.RoleAssignments).Include(u => u.Profile)
                .FirstOrDefaultAsync(u => u.Id == Guid.Parse(jwt.Subject), ct);
            if (user is null || user.Status == UserStatus.DISABLED) throw AppException.Unauthorized("Refresh token is invalid", "REFRESH_INVALID");
            return new AuthOutcome { Tokens = TokenPayload(user) };
        }
        catch (AppException) { throw; }
        catch { throw AppException.Unauthorized("Refresh token is invalid", "REFRESH_INVALID"); }
    }

    public async Task<object> MeAsync(Guid userId, CancellationToken ct)
    {
        var user = await _db.Users.Include(u => u.RoleAssignments).Include(u => u.Profile)
            .FirstOrDefaultAsync(u => u.Id == userId, ct) ?? throw AppException.Unauthorized();
        return SerializeUser(user);
    }

    public async Task ForgotAsync(string email, CancellationToken ct)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Email == email.Trim().ToLowerInvariant(), ct);
        if (user is null || user.Status == UserStatus.DISABLED) return;
        var token = Tokens.SignPurpose(_settings, user.Id, "password_reset", lifetime: "1h");
        await _email.SendAsync(
            user.Email,
            "Reset your ASAN Invest password",
            $"Open {_settings.ClientOrigin}/reset-password?token={Uri.EscapeDataString(token)} to choose a new password. This link expires in one hour.",
            ct);
    }

    public void RevokeRefresh(string? refreshToken)
    {
        if (string.IsNullOrWhiteSpace(refreshToken)) return;
        try
        {
            var jwt = Tokens.Require(refreshToken, _settings.JwtRefreshSecret, "refresh");
            var exp = jwt.ValidTo == DateTime.MinValue
                ? DateTimeOffset.UtcNow.Add(Tokens.ParseDuration(_settings.JwtRefreshExpiresIn))
                : new DateTimeOffset(DateTime.SpecifyKind(jwt.ValidTo, DateTimeKind.Utc));
            _challenges.RevokeRefresh(jwt.Id, exp);
        }
        catch
        {
            // Cookie delete still proceeds.
        }
    }

    public async Task ResetAsync(string token, string password, CancellationToken ct)
    {
        var jwt = Tokens.Require(token, _settings.JwtAccessSecret, "password_reset");
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == Guid.Parse(jwt.Subject), ct)
            ?? throw AppException.BadRequest("RESET_INVALID", "Reset token is invalid or expired");
        user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(password, _settings.BcryptRounds);
        await _db.SaveChangesAsync(ct);
    }

    public async Task VerifyEmailAsync(string token, CancellationToken ct)
    {
        var jwt = Tokens.Require(token, _settings.JwtAccessSecret, "email_verify");
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == Guid.Parse(jwt.Subject), ct)
            ?? throw AppException.BadRequest("VERIFY_INVALID", "Verification token is invalid or expired");
        user.EmailVerifiedAt = DateTimeOffset.UtcNow;
        user.IdentificationLevel = IdentificationLevel.BASIC;
        await _db.SaveChangesAsync(ct);
    }

    public static object AsanLogin() => new
    {
        provider = "asan_login",
        available = false,
        identificationLevelIfCompleted = "LEGAL",
        message = "ASAN Login is not connected in Phase 1. Use email registration (level 1) and continue legal actions via a representative or when the provider is specified.",
    };

    public async Task<object> CreateGuestAsync(CancellationToken ct)
    {
        var session = new GuestSession { ExpiresAt = DateTimeOffset.UtcNow.AddDays(7) };
        _db.GuestSessions.Add(session);
        await _db.SaveChangesAsync(ct);
        return new { token = session.Id, expiresInDays = 7 };
    }

    public async Task<object> PatchGuestAsync(Guid id, JsonElement answers, string? email, string? locale, CancellationToken ct)
    {
        var session = await _db.GuestSessions.FirstOrDefaultAsync(s => s.Id == id, ct) ?? throw AppException.NotFound();
        if (session.ExpiresAt <= DateTimeOffset.UtcNow || session.ConvertedUserId is not null) throw AppException.NotFound();
        session.Answers = answers.GetRawText();
        if (email is not null) session.Email = email.ToLowerInvariant();
        if (locale is not null) session.Locale = locale;
        await _db.SaveChangesAsync(ct);
        return new { token = session.Id, answers = JsonSerializer.Deserialize<object>(session.Answers) };
    }

    private static string NewOtp()
    {
        var value = RandomNumberGenerator.GetInt32(100000, 1000000);
        return value.ToString();
    }

    private TokenResult TokenPayload(User user)
    {
        var roles = Roles.Active(user.RoleAssignments, DateTimeOffset.UtcNow);
        var access = Tokens.SignAccess(_settings, user.Id, user.Email, roles, user.IdentificationLevel);
        var refresh = Tokens.SignRefresh(_settings, user.Id);
        var expires = DateTimeOffset.UtcNow.Add(Tokens.ParseDuration(_settings.JwtRefreshExpiresIn));
        return new TokenResult(access, refresh, expires, SerializeUser(user));
    }

    public static object SerializeUser(User user)
    {
        var roles = Roles.Active(user.RoleAssignments, DateTimeOffset.UtcNow);
        return new
        {
            id = user.Id,
            email = user.Email,
            roles,
            identificationLevel = user.IdentificationLevel,
            locale = user.Locale,
            status = user.Status,
            authProvider = user.AuthProvider,
            emailVerifiedAt = user.EmailVerifiedAt,
            twoFactorEnabled = user.TwoFactorEnabled,
            institutionId = user.InstitutionId,
            consents = JsonSerializer.Deserialize<object>(user.Consents),
            consentVersion = user.ConsentVersion,
            consentedAt = user.ConsentedAt,
            pepSanctionsStatus = user.PepSanctionsStatus,
            pepSanctionsCheckedAt = user.PepSanctionsCheckedAt,
            profile = user.Profile is null ? null : SerializeProfile(user.Profile),
        };
    }

    public static object SerializeProfile(Profile p) => new
    {
        id = p.Id,
        countryId = p.CountryId,
        sectorId = p.SectorId,
        activityAreaId = p.ActivityAreaId,
        contacts = JsonSerializer.Deserialize<object>(p.Contacts),
        companyName = p.CompanyName,
        companyCountryId = p.CompanyCountryId,
        companyRegId = p.CompanyRegId,
        taxId = p.TaxId,
        companyActivity = p.CompanyActivity,
        uboStructure = p.UboStructure is null ? null : JsonSerializer.Deserialize<object>(p.UboStructure),
        version = p.Version,
    };

    private async Task Audit(Guid? actor, string action, string objectType, string objectId, CancellationToken ct)
    {
        _db.AuditRecords.Add(new AuditRecord { ActorUserId = actor, Action = action, ObjectType = objectType, ObjectId = objectId });
        await _db.SaveChangesAsync(ct);
    }
}

public sealed record RegisterRequest(string Email, string Password, string? Locale, string? GuestSessionToken, ConsentBody? Consents);
public sealed record ConsentBody(string Version, bool PersonalData);

File: server/src/modules/auth/auth.service.ts
import type { User } from "@prisma/client";
import { env } from "../../config/env";
import { writeAudit } from "../../lib/audit";
import { AppError } from "../../lib/errors";
import { hashPassword, verifyPassword } from "../../lib/passwords";
import { prisma } from "../../lib/prisma";
import {
  generateNumericCode,
  generateOpaqueToken,
  hashToken,
  parseDurationToMs,
  signAccessToken,
} from "../../lib/tokens";
import { emailService } from "../../integrations/email/email.service";
import { requiresTwoFactor } from "../../domain/identification";
import { asanLoginStub } from "../../integrations/asan-imza/asan-imza.stub";

const INCLUDE = { profile: true } as const;

function issueAccess(user: User) {
  return signAccessToken({
    sub: user.id,
    roles: user.roles,
    email: user.email,
    identificationLevel: user.identificationLevel,
  });
}

async function persistRefreshToken(userId: string, userAgent?: string) {
  const token = generateOpaqueToken();
  const expiresAt = new Date(Date.now() + parseDurationToMs(env.JWT_REFRESH_EXPIRES_IN));
  await prisma.refreshToken.create({
    data: { userId, tokenHash: hashToken(token), expiresAt, userAgent: userAgent ?? null },
  });
  return { token, expiresAt };
}

async function copyGuestSession(userId: string, profileId: string, guestSessionToken?: string) {
  if (!guestSessionToken) return;
  const session = await prisma.guestSession.findUnique({ where: { tokenHash: hashToken(guestSessionToken) } });
  if (!session || session.expiresAt < new Date()) return;
  const answers = session.answers as Record<string, unknown>;
  await prisma.profile.update({
    where: { id: profileId },
    data: {
      guestAnswers: session.answers === null ? undefined : session.answers,
      country: (answers.country as string | undefined) ?? undefined,
      sector: (answers.sector as string | undefined) ?? undefined,
    },
  });
  await writeAudit({
    actorId: userId,
    action: "auth.guest_session_copied",
    objectType: "profile",
    objectId: profileId,
    after: { copied: true },
  });
}

export const authService = {
  async register(input: {
    email: string;
    password: string;
    locale?: string;
    guestSessionToken?: string;
    consents?: { version: string; personalData: true };
    ipAddress?: string;
  }) {
    const existing = await prisma.user.findUnique({ where: { email: input.email } });
    if (existing) throw AppError.conflict("An account with this email already exists", "EMAIL_TAKEN");

    const autoVerify = !env.REQUIRE_EMAIL_VERIFICATION;
    const user = await prisma.user.create({
      data: {
        email: input.email,
        passwordHash: await hashPassword(input.password),
        roles: ["investor"],
        locale: input.locale ?? "az",
        status: autoVerify ? "active" : "pending_verification",
        emailVerifiedAt: autoVerify ? new Date() : null,
        identificationLevel: autoVerify ? "basic" : "basic",
        consents: (input.consents ?? undefined) as object | undefined,
        consentVersion: input.consents?.version ?? null,
        consentedAt: input.consents ? new Date() : null,
        profile: { create: {} },
      },
      include: INCLUDE,
    });

    if (!autoVerify) {
      const token = generateOpaqueToken();
      await prisma.emailVerificationToken.create({
        data: { userId: user.id, tokenHash: hashToken(token), expiresAt: new Date(Date.now() + 24 * 60 * 60 * 1000) },
      });
      await emailService.send(user.email, "Verify your ASAN Invest account", `Verification token (logged only): ${token}`);
    }

    if (user.profile) {
      await copyGuestSession(user.id, user.profile.id, input.guestSessionToken);
    }

    await writeAudit({
      actorId: user.id,
      action: "auth.register",
      objectType: "user",
      objectId: user.id,
      ipAddress: input.ipAddress,
    });

    const accessToken = issueAccess(user);
    const refresh = await persistRefreshToken(user.id);
    return { user, accessToken, refresh };
  },

  async login(input: { email: string; password: string; userAgent?: string; ipAddress?: string }) {
    const user = await prisma.user.findUnique({ where: { email: input.email }, include: INCLUDE });
    if (!user || !user.passwordHash) {
      throw AppError.unauthorized("Invalid email or password", "INVALID_CREDENTIALS");
    }
    const matches = await verifyPassword(input.password, user.passwordHash);
    if (!matches) throw AppError.unauthorized("Invalid email or password", "INVALID_CREDENTIALS");
    if (user.status === "disabled") throw AppError.forbidden("Account is disabled", "ACCOUNT_DISABLED");
    if (user.status === "pending_verification" && env.REQUIRE_EMAIL_VERIFICATION) {
      throw AppError.forbidden("Email verification required", "EMAIL_NOT_VERIFIED");
    }

    if (requiresTwoFactor(user.roles)) {
      const code = env.NODE_ENV === "test" ? "123456" : generateNumericCode();
      const challenge = await prisma.twoFactorChallenge.create({
        data: {
          userId: user.id,
          codeHash: hashToken(code),
          expiresAt: new Date(Date.now() + 10 * 60 * 1000),
        },
      });
      await emailService.send(user.email, "ASAN Invest sign-in code", `Sign-in code (logged only): ${code}`);
      return { twoFactorRequired: true as const, challengeId: challenge.id };
    }

    const accessToken = issueAccess(user);
    const refresh = await persistRefreshToken(user.id, input.userAgent);
    await writeAudit({
      actorId: user.id,
      action: "auth.login",
      objectType: "user",
      objectId: user.id,
      ipAddress: input.ipAddress,
    });
    return { twoFactorRequired: false as const, user, accessToken, refresh };
  },

  async verifyTwoFactor(challengeId: string, code: string, userAgent?: string) {
    const challenge = await prisma.twoFactorChallenge.findUnique({
      where: { id: challengeId },
      include: { user: { include: INCLUDE } },
    });
    if (!challenge || challenge.usedAt || challenge.expiresAt < new Date()) {
      throw AppError.unauthorized("Two-factor challenge is invalid", "TWO_FACTOR_INVALID");
    }
    if (hashToken(code) !== challenge.codeHash) {
      throw AppError.unauthorized("Two-factor code is invalid", "TWO_FACTOR_INVALID");
    }
    await prisma.twoFactorChallenge.update({ where: { id: challenge.id }, data: { usedAt: new Date() } });
    if (!challenge.user.twoFactorEnabled) {
      await prisma.user.update({ where: { id: challenge.userId }, data: { twoFactorEnabled: true } });
    }
    const user = { ...challenge.user, twoFactorEnabled: true };
    const accessToken = issueAccess(user);
    const refresh = await persistRefreshToken(user.id, userAgent);
    return { user, accessToken, refresh };
  },

  async refresh(refreshToken: string | undefined) {
    if (!refreshToken) throw AppError.unauthorized("Refresh token missing", "REFRESH_MISSING");
    const stored = await prisma.refreshToken.findUnique({
      where: { tokenHash: hashToken(refreshToken) },
      include: { user: { include: INCLUDE } },
    });
    if (!stored || stored.revokedAt || stored.expiresAt < new Date()) {
      throw AppError.unauthorized("Refresh token is invalid", "REFRESH_INVALID");
    }
    await prisma.refreshToken.update({ where: { id: stored.id }, data: { revokedAt: new Date() } });
    const accessToken = issueAccess(stored.user);
    const refresh = await persistRefreshToken(stored.user.id, stored.userAgent ?? undefined);
    return { user: stored.user, accessToken, refresh };
  },

  async logout(refreshToken: string | undefined) {
    if (!refreshToken) return;
    await prisma.refreshToken.updateMany({
      where: { tokenHash: hashToken(refreshToken), revokedAt: null },
      data: { revokedAt: new Date() },
    });
  },

  async me(userId: string) {
    const user = await prisma.user.findUnique({ where: { id: userId }, include: INCLUDE });
    if (!user) throw AppError.unauthorized();
    return user;
  },

  async forgotPassword(email: string) {
    const user = await prisma.user.findUnique({ where: { email } });
    if (!user) return;
    const token = generateOpaqueToken();
    await prisma.passwordResetToken.create({
      data: { userId: user.id, tokenHash: hashToken(token), expiresAt: new Date(Date.now() + 60 * 60 * 1000) },
    });
    await emailService.send(user.email, "Reset your ASAN Invest password", `Reset token (logged only): ${token}`);
  },

  async resetPassword(token: string, password: string) {
    const stored = await prisma.passwordResetToken.findUnique({ where: { tokenHash: hashToken(token) } });
    if (!stored || stored.usedAt || stored.expiresAt < new Date()) {
      throw AppError.badRequest("RESET_INVALID", "Reset token is invalid or expired");
    }
    await prisma.$transaction([
      prisma.user.update({ where: { id: stored.userId }, data: { passwordHash: await hashPassword(password) } }),
      prisma.passwordResetToken.update({ where: { id: stored.id }, data: { usedAt: new Date() } }),
      prisma.refreshToken.updateMany({
        where: { userId: stored.userId, revokedAt: null },
        data: { revokedAt: new Date() },
      }),
    ]);
  },

  async verifyEmail(token: string) {
    const stored = await prisma.emailVerificationToken.findUnique({ where: { tokenHash: hashToken(token) } });
    if (!stored || stored.usedAt || stored.expiresAt < new Date()) {
      throw AppError.badRequest("VERIFY_INVALID", "Verification token is invalid or expired");
    }
    await prisma.$transaction([
      prisma.user.update({
        where: { id: stored.userId },
        data: { status: "active", emailVerifiedAt: new Date(), identificationLevel: "basic" },
      }),
      prisma.emailVerificationToken.update({ where: { id: stored.id }, data: { usedAt: new Date() } }),
    ]);
  },

  asanLogin() {
    return asanLoginStub.start();
  },
};
